File: Calculo Numerico/Trabalho 10/Trabalho10.cpp
/*
* Trabalho 10.
* Implemente o método de Simpson composto, 
* para calcular a integral de uma função do tipo:
* ln (a*x + (b*x)^(1/2) + (c*x)^(1/3) + (d*x)^(1/4))
*
* Calculo Numerico
* Claudio Schepke
* ---
* Ciencia da Computacao
* Universidade Federal do Pampa
* 
*
* Integrantes:
* ---
* Marcos V. Treviso, 121150107, marcosvtreviso@gmail.com
*
*
* Instrucoes:
* ---
* Compile: g++ -o trab Trabalho10.cpp -Wall -Wextra
* Run: ./trab
*
*/

#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <cstring>
#include <vector>

using namespace std;




const bool DEBUG = false;
double c1, c2, c3, c4; //a, b, c


double f(double x){
	return log((c1*x) + (c2*powf(x, (1/2.0))) + (c3*powf(x, (1/3.0))) + (c4*powf(x, (1/4.0))));
}


double metodoSimpsonComposto(int a, int b, int n){

	double h = (b-a)/(double) n;
	double soma = f(a) + f(b);
	
	for(int i=1; i<n; i++){
		if(i % 2 == 0)
			soma += 2 * f(a + i*h);
		else
			soma += 4 * f(a + i*h);
	}

	return  soma * h/3.0;
}




int main(){

	cout << fixed << setprecision(8);

	int a, b, n;
	while(true){
			
		cout << "Informe os valores dos coeficientes c1, c2, c3, c4:" << endl;
		cin >> c1 >> c2 >> c3 >> c4;

		cout << "Informe os intervalos de integracao a, b:" << endl;
		cin >> a >> b;

		cout << "Informe o numero de divisoes n:" << endl;
		cin >> n;

		cout << "Metodo de Simpson Composto: " << 
		metodoSimpsonComposto(a, b, n) << endl << endl;

	}

	return 0;
}

File: Calculo Numerico/Trabalho 10/Trabalho10.py
import math

def ln(x):
	return math.log(x, math.e)



a, b, c, d = 0, 0, 0, 0

def eq(x):
	return x*x - math.e**x


def getH(i, j, n):
	return (j-i)/(n*1.0)


def metodo_simpson_composta(i, j, n):
	
	h = getH(i, j, n)

	s = eq(i) + eq(j)
	for k in range(1, n):
		if k % 2 == 0:
			s += 2 * eq(i + k*h)
		else:
			s += 4 * eq(i + k*h)

	return  s * h/3.0




while(True):
	
	print("Digite os valores dos coeficientes a, b, c, d:")
	a, b, c, d = map(int, input().split())

	print("Digite os intervalos de integracao i, j:")
	i, j = map(int, input().split())

	if i <= 0 or j <= 0:
		print("Os valores dos intervalores devem ser maiores que 0")
		continue

	print("Digite o numero de divisoes n:")
	n = int(input())

	print("Metodo de Simpson Comp: %.8f" % metodo_simpson_composta(i, j, n))
	print("")

